pkg/hub/managedclusterset/global_managedclusterset_controller.go

```diff
@@ -0,0 +1,135 @@
+)
+
+const (
+	GlobalManagedClusterSetName = "global"
+)
+
+var GlobalManagedClusterSetSpec = clusterv1beta1.ManagedClusterSetSpec{
+	ClusterSelector: clusterv1beta1.ManagedClusterSelector{
+		SelectorType:  clusterv1beta1.LabelSelector,
+		LabelSelector: &metav1.LabelSelector{},
+	},
+}
+
+var globalManagedClusterSet = &clusterv1beta1.ManagedClusterSet{
+	ObjectMeta: metav1.ObjectMeta{
+		Name: GlobalManagedClusterSetName,
+	},
+	Spec: GlobalManagedClusterSetSpec,
+}
+
+type globalManagedClusterSetController struct {
+	clusterSetClient clustersetv1beta1.ClusterV1beta1Interface
+	clusterSetLister clusterlisterv1beta1.ManagedClusterSetLister
+	eventRecorder    events.Recorder
+}
+
+func NewGlobalManagedClusterSetController(
+	clusterSetClient clustersetv1beta1.ClusterV1beta1Interface,
+	clusterSetInformer clusterinformerv1beta1.ManagedClusterSetInformer,
+	recorder events.Recorder) factory.Controller {
+
+	c := &globalManagedClusterSetController{
+		clusterSetClient: clusterSetClient,
+		clusterSetLister: clusterSetInformer.Lister(),
+		eventRecorder:    recorder.WithComponentSuffix("global-managed-cluster-set-controller"),
+	}
+
+	return factory.New().
+		WithFilteredEventsInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			func(obj interface{}) bool {
+				metaObj, ok := obj.(metav1.ObjectMetaAccessor)
+				if !ok {
+					return false
+				}
+				// filter clustersets except globalManagedClusterSet.
+				return GlobalManagedClusterSetName != metaObj.GetObjectMeta().GetName()
+			},
+			clusterSetInformer.Informer(),
+		).
+		WithSync(c.sync).
+		// use ResyncEvery to make sure:
+		// 1. create the global clusterset once controller is launched
+		// 2. the global clusterset be recreated once it is deleted for some reason
+		ResyncEvery(10*time.Second).
+		ToController("GlobalManagedClusterSetController", recorder)
+}
+
+func (c *globalManagedClusterSetController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	globalClusterSetName := GlobalManagedClusterSetName
+	klog.V(4).Infof("Reconciling GlobalManagedClusterSet")
+
+	globalClusterSet, err := c.clusterSetLister.Get(globalClusterSetName)
+
+	// if the globalClusterSet not found, apply it.
+	if errors.IsNotFound(err) {
+		_, err := c.clusterSetClient.ManagedClusterSets().Create(ctx, globalManagedClusterSet, metav1.CreateOptions{})
+		if err == nil {
+			c.eventRecorder.Eventf("GlobalManagedClusterSetCreated", "Set the GlobalManagedClusterSet name to %+v. spec to %+v", globalClusterSetName, GlobalManagedClusterSetSpec)
+		}
+		return err
+
+	} else if err != nil {
+		return err
+	}
+
+	// if globalClusterSet is terminating, add the key to the controller queue with a second-long delay
+	if !globalClusterSet.DeletionTimestamp.IsZero() {
+		syncCtx.Queue().AddAfter(globalClusterSetName, 5*time.Second)
+		return nil
+	}
+
+	if err := c.syncGlobalClusterSet(ctx, globalClusterSet); err != nil {
```

Comment:
I would call it applyGlobalClusterSet and merge labels/annotation in it.

Reply:
Currently, we only update the global set spec.
And I also add a integration test case to handle it : https://github.com/open-cluster-management-io/registration/pull/251/files#diff-20b89e21483044eda78a48c9410652eacae6e58058adfb927b90dbece0af57a3R59